### PR TITLE
Add RTL propagation to HTML preprocessing

### PR DESCRIPTION
**Pull request by haroldtreen**
### Overview

Some articles have a text direction set using the `dir` property. When content extraction happens, sometimes the parent element where this is set is removed, thus reverting the text back to the default (`ltr`).

### Changes

- Adds a `propagateDirProperty` to `HtmlProcessor`. This will iterate over any passed selectors and propagate their `dir` down to their ancestor `<p>` tags. 

See: https://github.com/haroldtreen/epub-press/issues/10

### CODE REVIEW

File: lib/html-processor.js [hunk 1]
```diff
@@ -309,6 +309,16 @@ class HtmlProcessor {
 
         return HtmlProcessor.normalizeOutput(html, $);
     }
+
+    static propagateDirProperty(selector, html) {
```
Comment (seefood) [inline lib/html-processor.js]: OK, interesting approach... I hope it's enough to go just one level down. Is it already online to test or do I need to install a local server?

Comment (haroldtreen) [reply lib/html-processor.js]: Not out yet - I still need to test it out manually, but the integration + unit tests suggest this should work.

I used the BBC article you provided to test that the property would get sent down to all the children (not just one level). So I'm optimistic!

File: lib/content-extractor.js [hunk 2]
```diff
@@ -177,6 +177,7 @@ ContentExtractor.preprocess.OPERATIONS = {
         '.holder',
     ],
     mergeNodes: ['.article-text', '.section-inner', 'div.page', 'div.gtxt_body'],
+    propagateDirProperty: ['.direction'],
```
Comment (haroldtreen) [inline lib/content-extractor.js]: This is the class for the div where the BBC article puts their direction information - it probably won't be the same for all websites. I could also do something like "Check all the divs!" but wanted to start with the most lightweight approach and see how it aligns with other sites.

**Comment by seefood** [comment]
So how about a few more URLs? please pass them though this new code and attach the epub, and I'll go over them?

Thanks!

https://news.gooya.com/2018/10/post-20091.php
https://www.haaretz.co.il/literature/study/.premium-REVIEW-1.6588650
https://www.calcalist.co.il/local/articles/0,7340,L-3748731,00.html
http://www.aljazeera.net/news/international/2018/10/30/%D8%AA%D8%B1%D8%A7%D9%85%D8%A8-%D9%8A%D8%B1%D9%8A%D8%AF-%D8%A5%D9%86%D9%87%D8%A7%D8%A1-%D9%85%D9%86%D8%AD-%D8%A7%D9%84%D8%AC%D9%86%D8%B3%D9%8A%D8%A9-%D8%AA%D9%84%D9%82%D8%A7%D8%A6%D9%8A%D8%A7-%D9%84%D9%85%D9%88%D8%A7%D9%84%D9%8A%D8%AF-%D8%A3%D9%85%D9%8A%D8%B1%D9%83%D8%A7
https://ar.libyaobserver.ly/article/2261

**Comment by seefood** [comment]
oh yeah, do RTL and LTR articles mix OK in the same ePub?

**Comment by haroldtreen** [comment]
They should mix okay - only descendants of `rtl` nodes are marked. So if you had:

```html
<!-- Text Block 1 -->
<div dir="rtl">
   <p></p>
   <p></p>
</div>
<!-- Text Block 2 -->
<div>
   <p></p>
   <p></p>
</div>
```

Then only text block 1 would be impacted.

Thanks for the extra articles. Going to try and do a full test next.

**Comment by haroldtreen** [comment]
BCC Article should be fixed thanks to the above changes 👍 

image

**Comment by haroldtreen** [comment]
Nevermind! Got it fixed for all.

New approach:

- Extract the article
- Inspect all remaining paragraphs and what characters are in them.
- If the characters appear to be from a RTL language - add the direction property.

Going to merge and deploy this!

**Comment by seefood** [comment]
Oh cool! I didn't have the time to look at it al weekend :)

so it's now on the server? I'll go test!

**Pull request merged**